pkg/auth/ldaputil/query.go

```diff
@@ -0,0 +1,211 @@
+				TimeLimit:    config.TimeLimit,
+				Filter:       config.Filter,
+			},
+			QueryAttribute: config.QueryAttribute,
+		},
+		NameAttributes: nameAttributes,
+	}, nil
+}
+
+// NewSearchRequest creates a new search request from the identifying query by internalizing the value of
+// the attribute to be filtered as well as any additional attributest that need to be recovereds
+func (o *IdentifiyingLDAPQueryOptions) NewSearchRequest(attributeValue string,
+	additionalAttributes []string) (*ldap.SearchRequest, error) {
+
+	allAttributes := util.NewStringSet(o.NameAttributes...)
+	allAttributes.Insert(additionalAttributes...)
+
+	if o.QueryAttribute == "DN" || o.QueryAttribute == "dn" {
+		if !strings.Contains(attributeValue, o.BaseDN) {
+			return nil, fmt.Errorf("search for entry with %s=%s would search outside of the base dn specified (dn=%s)",
+				o.QueryAttribute, attributeValue, o.BaseDN)
+		}
+		if _, err := ldap.ParseDN(attributeValue); err != nil {
+			return nil, fmt.Errorf("could not search by dn, invalid dn value: %v", err)
+		}
+		return o.buildDNQuery(attributeValue, allAttributes.List()), nil
+	} else {
+		return o.buildAttributeQuery(attributeValue, allAttributes.List()), nil
+	}
+}
+
+// buildDNQuery builds the query that finds an LDAP entry with the given DN
+// this is done by setting the DN to be the base DN for the search and setting the search scope
+// to only consider the base object found
+func (o *IdentifiyingLDAPQueryOptions) buildDNQuery(dn string,
+	attributes []string) *ldap.SearchRequest {
+	return ldap.NewSearchRequest(
+		dn,
+		ldap.ScopeBaseObject, // over-ride original
+		int(o.DerefAliases),
+		0, // allowed return size - indicates no limit
+		o.TimeLimit,
+		false,           // not types only
+		"objectClass=*", // filter that returns all values
+		attributes,
+		nil, // no controls
+	)
+}
+
+// buildAttributeQuery builds the query containing a filter that conjoins the common filter given
+// in the configuration with the specific attribute filter for which the attribute value is given
+func (o *IdentifiyingLDAPQueryOptions) buildAttributeQuery(attributeValue string,
+	attributes []string) *ldap.SearchRequest {
+	specificFilter := fmt.Sprintf("%s=%s",
+		ldap.EscapeFilter(o.QueryAttribute),
+		ldap.EscapeFilter(attributeValue))
+
+	filter := fmt.Sprintf("(&(%s)(%s))", o.Filter, specificFilter)
+
+	return ldap.NewSearchRequest(
+		o.BaseDN,
+		int(o.Scope),
+		int(o.DerefAliases),
+		0, // allowed return size - indicates no limit
+		o.TimeLimit,
+		false, // not types only
+		filter,
+		attributes,
+		nil, // no controls
+	)
+}
+
+// QueryForUniqueEntry queries for an LDAP entry on an LDAP server determined from a ClientConfig
+// by creating a search request from the requisite input. The query is expected to return one unqiue
+// result. If this is not the case, errors are raised
+func QueryForUniqueEntry(clientConfig LDAPClientConfig,
+	query *ldap.SearchRequest) (entry *ldap.Entry, err error) {
```

Comment:
Don't name return values for function implementations.  Find them all in this pull and remove them.